vault/login_mfa.go

```diff
@@ -0,0 +1,2798 @@
+package vault
+
+import (
+	"bytes"
+	"context"
+	"encoding/base64"
+	"errors"
+	"fmt"
+	"image/png"
+	"io/ioutil"
+	"net/http"
+	"net/url"
+	"path"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/chrismalek/oktasdk-go/okta"
+	"github.com/dgrijalva/jwt-go"
+	duoapi "github.com/duosecurity/duo_api_golang"
+	"github.com/duosecurity/duo_api_golang/authapi"
+	"github.com/golang/protobuf/proto"
+	"github.com/hashicorp/errwrap"
+	"github.com/hashicorp/go-cleanhttp"
+	"github.com/hashicorp/go-hclog"
+	"github.com/hashicorp/go-memdb"
+	"github.com/hashicorp/go-multierror"
+	"github.com/hashicorp/go-uuid"
+	"github.com/hashicorp/vault-licensing/license"
+	"github.com/hashicorp/vault/helper/identity"
+	"github.com/hashicorp/vault/helper/identity/mfa"
+	"github.com/hashicorp/vault/helper/namespace"
+	"github.com/hashicorp/vault/sdk/framework"
+	"github.com/hashicorp/vault/sdk/helper/jsonutil"
+	osslicense "github.com/hashicorp/vault/sdk/helper/license"
+	"github.com/hashicorp/vault/sdk/helper/parseutil"
+	"github.com/hashicorp/vault/sdk/helper/strutil"
+	"github.com/hashicorp/vault/sdk/logical"
+	"github.com/hashicorp/vault/vault/quotas"
+	"github.com/mitchellh/mapstructure"
+	otplib "github.com/pquerna/otp"
+	totplib "github.com/pquerna/otp/totp"
+)
+
+const (
+	mfaMethodTypeTOTP              = "totp"
+	mfaMethodTypeDuo               = "duo"
+	mfaMethodTypeOkta              = "okta"
+	mfaMethodTypePingID            = "pingid"
+	memDBLoginMFAConfigsTable      = "login_mfa_configs"
+	memDBMFALoginEnforcementsTable = "login_enforcements"
+	mfaTOTPKeysPrefix              = systemBarrierPrefix + "mfa/totpkeys/"
+
+	// loginMFAConfigPrefix is the storage prefix for persisting login MFA method
+	// configs
+	loginMFAConfigPrefix      = "login-mfa/method/"
+	mfaLoginEnforcementPrefix = "login-mfa/enforcement/"
+)
+
+type totpKey struct {
+	Key string `json:"key"`
+}
+
+// loginMfaPaths returns the API endpoints to configure the new style
+// login MFA. The following paths are supported:
+// mfa/method-id/:mfa_method - management of MFA method IDs, which can be used for configuration
+// mfa/login_enforcement/:config_name - configures single or two phase MFA auth
+func (b *SystemBackend) loginMFAPaths() []*framework.Path {
```

Comment:
This is not hooked right now as far as I could tell. Should it be added here? https://github.com/hashicorp/vault/blob/23301d73431609dd6ffaf4fc55cd3932b1c7578c/vault/logical_system.go#L163-L187